sdk/cxx/include/sawtooth/transaction_handler.h

```diff
@@ -0,0 +1,111 @@
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+------------------------------------------------------------------------------*/
+#pragma once
+
+#include <map>
+#include <memory>
+#include <list>
+#include <string>
+
+#include "proto/transaction.pb.h"
+
+#include "sawtooth/global_state.h"
+
+namespace sawtooth {
+
+// Utility types to allow for managed instances of std::string on the heap.
+typedef std::unique_ptr<std::string> StringUPtr;
+typedef std::shared_ptr<std::string> StringPtr;
+
+// The transaction data for a Transaction Processing request.
+class Transaction {
+ public:
+    Transaction(StringPtr header, StringPtr payload, StringPtr signature):
+            payload_(payload), signature_(signature) {
+        this->header_.ParseFromArray(header->c_str(),
+            header->length());
+    }
+    virtual ~Transaction() { }
+
+    const TransactionHeader& header() const {
+        return this->header_;
+    }
+
+    const std::string& payload() const {
+        return *(this->payload_);
+    }
+
+    const std::string& signature() const {
+        return *(this->signature_);
+    }
+
+ private:
+    TransactionHeader header_;
+    StringPtr payload_;
+    StringPtr signature_;
+};
+typedef std::unique_ptr<Transaction> TransactionUPtr;
+
+
+// TransactionApplicator is an interface that defined to represent the
+// processing of a single transaction. This object should make no assumptions
+// about the thread and timing of when it will execute. The
+// TransactionApplicator owns the reference to the transaction and the global
+// state it is passed.
+class TransactionApplicator {
+ public:
+    TransactionApplicator(sawtooth::TransactionUPtr txn,
+        sawtooth::GlobalStateUPtr state) : txn(std::move(txn)),
+        state(std::move(state)) { }
+    // Process the transaction, this function should do all of the processing
+    // of the given transaction. The general framework should be:
+    // 1) Validate the transaction
+    // 2) retrieve the state from GlobalState
+    // 3) validate the transaction changes are valid
+    // 4) Write the updated state to GlobalState
+    //
+    // Any exception thrown from this function will result in the transaction
+    // being marked as invalid.
+    virtual void Apply() = 0;
+ protected:
+    sawtooth::TransactionUPtr txn;
+    sawtooth::GlobalStateUPtr state;
+};
+typedef std::unique_ptr<TransactionApplicator> TransactionApplicatorUPtr;
+
+
+// Definition of a TransactionHandler to be registered with the validator.
+// Every Transaction Processor should register at least one of these. This
+// definition allows for the handler to declare multiple version and encodings
+// that it supports.
+class TransactionHandler {
```

Comment:
To enable parallel processing of Transactions ( not enabled yet in the c++ sdk, just a thread-pool away). Without the restriction that the Apply method being re-entrant.